test/statement/for_in.stmt

```diff
@@ -58,6 +58,13 @@ for (var identifier
     in iteratableExpression) {
   body;
 }
+>>> Split inside initializer.
+for (var identifier in someVeryLong + iterableExpression) { body; }
+<<<
+for (var identifier in someVeryLong +
+    iterableExpression) {
```

Comment:
Is this what you meant by `newline in the initializer of a for-in loop forced the in to split`?
Honestly, I thought the split was going to be 
```
for (var identifier in
     someVeryLong + iterableExpression) {
  body;
}
```

Reply:
There are two separate style choices:

1. If a newline occurs inside the expression after `in`, does that force us to split before the `in` as well?

In many places, a newline inside a subexpression will also force the outer expression to split. For example, the formatter will never produce:

```dart
expression1 + (expression2 *
    expression3);
```

If either operand to an infix operator contains a newline, we force the infix operator to split too (even if doing so isn't necessary to fit in the page width), so you get:

```dart
expression1 +
    (expression2 * expression3);
```

Or:

```dart
expression1 +
    (expression2 *
        expression3);
```

Depending on whether splitting at the outer infix operator is enough to not need to also split the inner infix operator.

For "assignment" pieces (`=`, `=>`, `in`, `:`), there is similar logic. For example, if you have a really long map key or value (we use AssignPiece for map entries too), you can get:

```dart
var map = {
  someReallyLong + keyExpression:
    theValue
}
```

But not:

```dart
var map = {
  someReallyLong +
      keyExpression: theValue
}
```

Before this PR, `in` had the same behavior, which meant you could get:

```dart
for (var identifier
      in someVeryLong +
          iterableExpression) {
  body;
}
```

Or:

```dart
for (var identifier
      in someVeryLong + iterableExpression) {
  body;
}
```

But never:

```dart
for (var identifier in someVeryLong +
    iterableExpression) {
  body;
}
```

Whether we want to allow the latter is basically a style choice. The old formatting style tries to be pretty dense and avoid newlines when possible, but my impression is that users generally like a sort of taller, more open style. So when I started working on the new style, one of the changes I made is that for assignment-like things, it generally does have the infix-operator-like rule that a split inside an operand forces the assignment to split.

But I think that choice looks pretty funny for `in`. Honestly, I'm not sure it's the right fit for `=` either. I think many users would be OK with:

```dart
variable = someLongExpression +
    thatWraps;
```

So it may be that we end up changing other assignment-like pieces to allow splitting in the RHS without forcing the assignment operator to split. But, for now, I just did it for `in`. 

There's a long TODO(tall) comment in AssignPiece about this so that our future selves can play with that style choice once we can run the formatter on a corpus.

2. If the whole for clause doesn't fit on one line because the initializer is too long, do we prefer to split before the `in`, or inside the initializer itself?

Assuming that we do allow a split inside the RHS after `in` (style choice 1) and also allow splitting before the `in`, there is a separate style choice for which of those should be *preferred* when either is sufficient to fit in the page width. I think that's the question you're asking.

In general, the formatter usually prefer to split at outer constructs before inner ones. In fact, one of the main ways it models that preference is the above style choice where in many contexts, splitting in a subexpression *forces* the outer one to split. That leads the formatter to only be able to choose between "split at outer and not inner" and "split at *both* outer and inner". Since fewer splits are always better, it will always then prefer to split at the outer expression and not the inner if that fits.

But in places where we don't have that hard constraint (like with `in` here), we now have a choice as to which should be preferred.

I don't have a strong sense of what the right answer is. We'll probably want to iterate on it. For what it's worth, the PR here follows what the old style does and I can't recall people complaining about how newlines for-in loop initializers get handled. I added a TODO comment so that we reconsider it when things are further along.